firebase-functions/src/api/v1/routes/user/manage-permissions.ts

```diff
@@ -0,0 +1,117 @@
+import admin from "firebase-admin";
+import type { Request, Response } from "express";
+import { PermissionError, ValidationError, UnknownError } from "../../../error";
+import { PERMISSIONS } from "../../../../shared/constants";
+
+export default async function managePermissions(req: Request, res: Response) {
+  const adminUserId = req.userId;
+  const adminPermissions = req.userPermissions;
+  const userId = req.params.userId;
+  const permissions = req.body.permissions as Record<
+    keyof typeof PERMISSIONS,
+    boolean
+  >;
+
+  if (!adminPermissions.manageUserPermissions) {
+    res
+      .status(403)
+      .json(
+        new PermissionError(
+          "You do not have permission to manage another user's permissions"
+        )
+      );
+    return;
+  }
+
+  const permissionKeys = Object.keys(permissions || {});
+
+  if (permissionKeys.length === 0) {
+    res
+      .status(400)
+      .json(new ValidationError("Must provide permissions in post body."));
+    return;
+  }
+
+  // provisioned is the special permission we use to indicate that
+  // a user account is fully set up, so it should not be modified
+  // in any way outside of the provision route
+  if ("provisioned" in permissions) {
+    res
+      .status(400)
+      .json(new ValidationError("Cannot change provisioned permission."));
+    return;
+  }
+
+  // prevent an admin user from removing the manageUserPermissions option
+  // from self. This precents us from accidentally getting into the circumstance
+  // where the only admin revokes the ability to manage user permissions
+  // and there's no one else that is able to do it
+  if (adminUserId === userId && "manageUserPermissions" in permissions) {
+    res
+      .status(400)
+      .json(
+        new ValidationError(
+          "You cannot change the manage user permissions option for yourself. Enlist another user with the `manage user permissions` permission to do this for you."
+        )
+      );
+    return;
+  }
+
+  const invalidPermissions = permissionKeys.filter((key) => {
+    return !(key in PERMISSIONS);
+  });
+
+  if (invalidPermissions.length > 0) {
+    res
+      .status(400)
+      .json(
+        new ValidationError(
+          `Invalid permission(s): ${invalidPermissions.join(", ")}`
+        )
+      );
+    return;
+  }
```

Comment:
Everything above this is just validating the request. Is this something that should be in middleware instead of a part of this handler? I'm not sure if we'd reuse this logic, but if so that's something to consider.

Reply:
This should be the only route that does these kinds of validations, so I'm inclined to keep it here. If I'm wrong, and in the future, we have multiple places doing stuff like this, it would be fine to pull out into middleware.